src/pyk/proof/exploration.py

```diff
@@ -0,0 +1,359 @@
+
+        super().__init__(id, proof_dir=proof_dir, subproof_ids=subproof_ids, admitted=admitted)
+        self.kcfg = kcfg
+        self.init = kcfg._resolve(init)
+        self.logs = logs
+
+        self._terminal_node_ids = {
+            kcfg._resolve(terminal_node_id)
+            for terminal_node_id in (terminal_node_ids if terminal_node_ids is not None else [])
+        }
+
+        self.kcfg.cfg_dir = self.proof_subdir / 'kcfg' if self.proof_subdir else None
+
+        if self.proof_dir is not None and self.proof_subdir is not None:
+            ensure_dir_path(self.proof_dir)
+            ensure_dir_path(self.proof_subdir)
+
+    def is_init(self, node_id: NodeIdLike) -> bool:
+        return self.kcfg._resolve(node_id) == self.init
+
+    def is_terminal(self, node_id: NodeIdLike) -> bool:
+        return self.kcfg._resolve(node_id) in self._terminal_node_ids
+
+    def is_pending(self, node_id: NodeIdLike) -> bool:
+        return self.kcfg.is_leaf(node_id) and not self.is_terminal(node_id) and not self.kcfg.is_stuck(node_id)
+
+    @property
+    def pending(self) -> list[KCFG.Node]:
+        return [nd for nd in self.kcfg.leaves if self.is_pending(nd.id)]
+
+    @property
+    def failing(self) -> list[KCFG.Node]:
+        return []
+
+    def add_terminal(self, node_id: NodeIdLike) -> None:
+        self._terminal_node_ids.add(self.kcfg._resolve(node_id))
+
+    def remove_terminal(self, node_id: NodeIdLike) -> None:
+        node_id = self.kcfg._resolve(node_id)
+        if node_id not in self._terminal_node_ids:
+            raise ValueError(f'Node is not terminal: {node_id}')
+        self._terminal_node_ids.remove(node_id)
+
+    def prune_from(self, node_id: NodeIdLike, keep_nodes: Iterable[NodeIdLike]) -> list[NodeIdLike]:
+        pruned_nodes = self.kcfg.prune(node_id, keep_nodes=list(keep_nodes) + [self.init])
+        for nid in pruned_nodes:
+            self._terminal_node_ids.discard(self.kcfg._resolve(nid))
+        return pruned_nodes
+
+    def shortest_path_to(self, node_id: NodeIdLike) -> tuple[KCFG.Successor, ...]:
+        spb = self.kcfg.shortest_path_between(self.init, node_id)
+        assert spb is not None
+        return spb
+
+    @staticmethod
+    def read_proof(id: str, proof_dir: Path) -> ExplorationProof:
+        proof_path = proof_dir / f'{hash_str(id)}.json'
+        if ExplorationProof.proof_exists(id, proof_dir):
+            proof_dict = json.loads(proof_path.read_text())
+            _LOGGER.info(f'Reading ExplorationProof from file {id}: {proof_path}')
+            return ExplorationProof.from_dict(proof_dict, proof_dir=proof_dir)
+        raise ValueError(f'Could not load ExplorationProof from file {id}: {proof_path}')
+
+    @property
+    def status(self) -> ProofStatus:
+        if len(self.kcfg.stuck) > 0:
+            return ProofStatus.FAILED
+        elif len(self.pending) > 0:
+            return ProofStatus.PENDING
+        else:
+            return ProofStatus.PASSED
```

Comment:
```suggestion
    @property
    def status(self) -> ProofStatus:
        if len(self.kcfg.stuck) > 0 or self.subproofs_status == ProofStatus.FAILED:
            return ProofStatus.FAILED
        elif len(self.pending) > 0 or self.subproofs_status == ProofStatus.PENDING:
            return ProofStatus.PENDING
        else:
            return ProofStatus.PASSED
```

Does it really make sense to have a `FAILED` result here, even for `stuck` nodes? Maybe this should only return `PENDING` or `PASSED`.

But this makes me think that this class structure just doesn't make sense. This may be better suited as just a method added to `KCFGExplore` which takes in a `KCFGSemantics` as well.

Reply:
If I remember correctly, in one discussion you said something along the lines of "`stuck` nodes should not be happening in a correctly-defined semantics". I am following that intuition to mark a proof with stuck nodes as failed.

`ExplorationProof`s are currently, by construction, not allowed to have `subproofs`. What would a subproof mean in an `ExplorationProof`?